[OCPNODE-553] bump(*) and config MirrorByDigestOnly using ImageContentPolicy

QiWang19 opened the pull request:
Replace ImageContentSourcePolicy with config/v1 ImageContentPolicy. 
Support using mirror by tags by setting MirrorByDigestOnly value using `AllowMirrorByTags` field from ImageContentPolicy instead of hardcode `true`.

- MCO PR support allowMirrorByTags: https://github.com/openshift/machine-config-operator/pull/2755
- ImageContentPolicy API: https://github.com/openshift/api/pull/874
- Story: https://issues.redhat.com/browse/OCPNODE-553
- REF: https://issues.redhat.com/browse/OCPNODE-521

Signed-off-by: Qi Wang <qiwan@redhat.com>

adambkaplan commented on go.mod:
```diff
@@ -6,7 +6,9 @@ require (
+	github.com/openshift/build-machinery-go v0.0.0-20210712174854-1bb7fd1518d3
 	github.com/stretchr/testify v1.7.0
-	k8s.io/apimachinery v0.20.1
+	k8s.io/apimachinery v0.22.1
```
`go.mod` changes should be captured in a separate commit, with the title like "bump(x):"

In this case, I would use the commit title "bump(*):", with a commit message along the lines of "Updating apimachinery to v0.22.1 and build-machinery-go to latest"

adambkaplan commented on go.mod:
```diff
@@ -6,7 +6,9 @@ require (
+	k8s.io/apimachinery v0.22.1
 )
+
+replace github.com/openshift/api => github.com/QiWang19/api v0.0.0-20210913190153-b460ec7c7527
```
In your final version you will want to remove this and bump openshift-api

adambkaplan commented on pkg/registries/registries.go:
```diff
@@ -159,7 +159,7 @@ func EditRegistriesConfig(config *sysregistriesv2.V2RegistriesConf, insecureScop
 	for _, mirrorSet := range mirrorSets {
 		reg := getRegistryEntry(mirrorSet.Source)
-		reg.MirrorByDigestOnly = true
+		reg.MirrorByDigestOnly = mirrorSet.AllowMirrorByTags
```
Shouldn't this be a "not"?
```suggestion
		reg.MirrorByDigestOnly = !mirrorSet.AllowMirrorByTags
```

adambkaplan reviewed (commented):
A few comments here to guide your pull request.

As this is library code, this PR also needs unit tests to verify the desired behavior.

QiWang19 commented:
> A few comments here to guide your pull request.
> 
> As this is library code, this PR also needs unit tests to verify the desired behavior.

@adambkaplan  Thanks for the guidance. The original github.com/openshift/api that this PR relies on has been reworked. I will update based on the new changes.

QiWang19 committed:
bump(*): update openshift/api to master for CRD imagecontentpolicy

Bump openshift/api to https://github.com/openshift/api/commit/636513e97fdaf2dc65efd95e8c239f892057f791
to import new CRD imagecontentpolicy.

Signed-off-by: Qi Wang <qiwan@redhat.com>

QiWang19 committed:
Replace ICSP with apicfgv1.ImageContentPolicy

Replace operatore/v1alpha1 ImageContentSourcePolicy structs with config/v1 ImageContentPolicy.

Signed-off-by: Qi Wang <qiwan@redhat.com>

QiWang19 commented:
@adambkaplan Could you review? Not sure how I can fix test failure 
```
vendor/github.com/openshift/build-machinery-go/make/targets/golang/../../lib/golang.mk:22: *** `go` is required with minimal version "1.15.2", detected version "1.13.8". You can override this check by using `make GO_REQUIRED_MIN_VERSION:=`.  Stop.
{"component":"entrypoint","error":"wrapped process failed: exit status 2","file":"prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2021-09-30T18:13:16Z"} 
```

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -71,24 +71,29 @@ func mergedMirrorSets(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy
 		ds := disjointSets[source]
 		topoGraph := newTopoGraph()
 		for _, set := range *ds {
+			// AllowMirrorByTags is only set once for each source in imagecontentpolicies.config.openshift.io
```
[Hum, so the external API doesn’t match the capabilities of the underlying implementation 😢 ]

Just wondering, would it make sense to fail, treating conflicting configuration as invalid? I’m _very weakly_ leaning in that direction, but I can live with the above, and there might well be something I’m missing.

QiWang19 replied:
The external [API](https://github.com/openshift/api/blob/master/config/v1/types_image_content_policy.go#L46) added the kubebuilder validation to use source as key of the map since I saw the mirrors are merged in the mergedMirrorSets() use the source as the key of the map. We can change that API since it's a new CRD we want to add.
If enables setting multiple same source, how should we handle the conflict `allowMirrorByTags` for the same source? e.g. allowMirrorByTags is `true` for source:A, mirrors:[mirror0, mirror1, mirror2], and `false` for source:A, mirrors: [mirror1, mirror2]. Do we still use current logic to merge the mirrors by the source? @mtrmac WDYT?

mtrmac replied:
> The external [API](https://github.com/openshift/api/blob/master/config/v1/types_image_content_policy.go#L46) added the kubebuilder validation to use source as key of the map since I saw the mirrors are merged in the mergedMirrorSets() use the source as the key of the map. We can change that API since it's a new CRD we want to add.

I can’t find a precise definition of the `listType=map` annotation, but if it means that the data should be expressible as a `map[source]…`, with each `source` value being used at most once, that looks, from a historical point of view at least, incorrect, or at least insufficient.

(If we wanted to enforce uniqueness, the code in this subpackage could have just refused to work on duplicates, instead of doing all of the sorting and merging that it currently does.)

On the contrary, https://github.com/openshift/api/blob/636513e97fdaf2dc65efd95e8c239f892057f791/config/v1/types_image_content_policy.go#L39-L43 _explicitly_ documents that multiple entries with the same “source” is a valid configuration, and prescribes the kind of merging this package does.

(This means that an administrator or an operator can always add an ICSP CR object, without manually merging with any existing data, and leaving the merging work to be done by MCO or other consumers.)

As yet another argument, even if the `ICP.Spec.RepositoryDigestMirrors` field were defined as a map with unique `source` values, AFAICS MCO is listing _all_ (currently) ICSP resources and merges the data (note that the input of `mergedMirrorSets` is an _array_ of `[]*apicfgv1.ImageContentPolicy`; so unique `source` values within one CR doesn’t mean anything WRT uniqueness across CRs.

So if we can still change the CRD, I’d prefer the map-like behavior annotation to be dropped (if that’s what the annotation is).

---

Actually, if we can still change the CRD, shouldn’t `RepositoryDigestMirrors` be just `RepositoryMirrors`, now that they are not always digest-restricted?

---

> If enables setting multiple same source, how should we handle the conflict `allowMirrorByTags` for the same source? e.g. allowMirrorByTags is `true` for source:A, mirrors:[mirror0, mirror1, mirror2], and `false` for source:A, mirrors: [mirror1, mirror2]. Do we still use current logic to merge the mirrors by the source? @mtrmac WDYT?

Yes, that’s the problem I was asking about.

We can’t just _not merge_ the entries; `registries.conf` only uses one value per a prefix, so creating a config file that has two entries for the same `source` would just not represent some of the mirrors.

The technically correct/clean implementation, given the current CRD, would be to reject configurations that specify conflicting values; that either does exactly what the user asked, or refuses to do anything.

OTOH that also means that the “just add an ICSP CR” use case described above no longer works that easily, because the newly added CR would need to have an `AllowMirrorByTags` value that matches any pre-existing CRs.

Hence I don’t have a strong opinion.

It’s generally easier to be more restrictive and loosen restrictions, than to ship a less restrictive implementation and tighten it later (possibly breaking users), so I’m leaning towards the stricter approach of rejecting conflicts, but that’s not a strong argument and there could well be a good reason not to do that.

QiWang19 replied:
> On the contrary, https://github.com/openshift/api/blob/636513e97fdaf2dc65efd95e8c239f892057f791/config/v1/types_image_content_policy.go#L39-L43 explicitly documents that multiple entries with the same “source” is a valid configuration, and prescribes the kind of merging this package does. 

The `multiple entries with the same “source” is a valid` was a mistake. The listType=map will limit the same `source` can only be used once. I don't think the openshift/api reviewers will disallow the changes on this CRD since it was just newly created for the implementation of allowMirrorByTags. I will create a new PR and the comment can leave it as is if the listType=map is removed.

> Actually, if we can still change the CRD, shouldn’t RepositoryDigestMirrors be just RepositoryMirrors, now that they are not always digest-restricted?

AllowMirrorByTags makes the mirror-by-tags only configurable but does not change the current default setting. The digest restriction still exists as a default. Should we just keep the Digest in the name RepositoryDigestMirrors?

> The technically correct/clean implementation, given the current CRD, would be to reject configurations that specify conflicting values; that either does exactly what the user asked, or refuses to do anything.
> 
> OTOH that also means that the “just add an ICSP CR” use case described above no longer works that easily, because the newly added CR would need to have an AllowMirrorByTags value that matches any pre-existing CRs.

I am confused about the definitions of conflict configurations. So the rule of rejecting conflicting values would be, going through the list of  []*apicfgv1.ImageContentPolicy, if `allowMirrorByTags` value is different for the same source, we think of it as conflict?  Although the user may just want to update the `allowMirrorByTags`, we have no idea which value is the user expected.
Yes, the “just add an ICSP CR” use case is not that easy. When the `allowMirrorByTags` value needs to change for the source that exists in pre-exising CR, the existed CRs should be deleted before applying a new one in order to avoid the conflict.

mtrmac replied:
> I will create a new PR and the comment can leave it as is if the listType=map is removed.

I’m a bit confused, just to confirm; do I understand it correctly that you are proposing to remove the `listType=map` and not the comment?

---

> Should we just keep the Digest in the name RepositoryDigestMirrors?

I think it would be a bit cleaner without the `Digest`, OTOH it’s more renaming for users to deal with. I very weakly prefer removing it, but I’m fine with it staying as is.

---

> So the rule of rejecting conflicting values would be, going through the list of []*apicfgv1.ImageContentPolicy, if `allowMirrorByTags` value is different for the same source, we think of it as conflict? Although the user may just want to update the `allowMirrorByTags`, we have no idea which value is the user expected.

Yes, that’s what I mean by a conflict.

(Another option would be to define the CR so that `allowMirrorByTags` is an optional value, if CRDs can express an optional boolean: then users who need to define this value could define one, and users who just want to add a mirror without setting policy could add mirrors without including `allowMirrorByTags` in their CR. OTOH that’s making it even more complex.)

mtrmac replied:
> and not the comment?

… saying that multiple entries with the same “source” is valid.

QiWang19 replied:
> I’m a bit confused, just to confirm; do I understand it correctly that you are proposing to remove the `listType=map` and not the comment?

@mtrmac  Yes. Keep the comment about `multiple same source`, and remove the `listType=map` validation.

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -71,24 +71,29 @@ func mergedMirrorSets(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy
 		topoGraph := newTopoGraph()
 		for _, set := range *ds {
+			// AllowMirrorByTags is only set once for each source in imagecontentpolicies.config.openshift.io
+			if !allowMirrorByTagsFlag && set.AllowMirrorByTags {
```
(I don’t think the `!allowMirrorByTagsFlag &&` part is necessary, but it doesn’t hurt.)

mtrmac commented on pkg/registries/registries.go:
```diff
@@ -71,24 +71,29 @@ func mergedMirrorSets(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy
-	res := []apioperatorsv1alpha1.RepositoryDigestMirrors{}
+	res := []apicfgv1.RepositoryDigestMirrors{}
 	for _, source := range sources {
+		allowMirrorByTagsFlag := false
```
(Maybe `s/Flag//` ? “allow…” feels sufficient as a name of a boolean to me.)

mtrmac commented on pkg/registries/registries_test.go:
```diff
@@ -297,13 +297,16 @@ func TestEditRegistriesConfig(t *testing.T) {
 			name:     "insecure+blocked prefixes with wildcard entries",
```
Conceptually I’d prefer a separate test case for the new “allow tags” merging implementation, clearly targeted at that aspect, and with a name to match. Right now, it’s not obvious that this test case is intended to do that.

mtrmac reviewed (commented):
ACK to the overall idea/approach.

adambkaplan commented:
/hold

Per the most recent status in JIRA, there were API concerns which needed to be reworked [1] @QiWang19 has any progress been made since?

[1] https://issues.redhat.com/browse/BUILD-291?focusedCommentId=19181313&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-19181313

adambkaplan commented:
bump @QiWang19 - any progress on the API for this?

QiWang19 commented:
No progress

QiWang19 commented:
Close since the external API will be changed openshift/enhancements#929
